Flux query basics (#2887)
influxdata/docs-v2: content/flux/v0.x/get-started/query-basics.md

---
title: Flux query basics
description: >
  View basic concepts and steps to use when querying data with Flux. 
menu:
  flux_0_x:
    name: Query basics
    parent: Get started
weight: 102
---

Most Flux queries follow the same basic structure.
Familiarize yourself with the basic concepts and steps to use when querying data with Flux. 

- [Important concepts](#important-concepts)
- [Basic query structure](#basic-query-structure)
- [Write a basic query](#write-a-basic-query)

## Important concepts

- [Pipe-forward operator](#pipe-forward-operator)
- [Predicate expressions](#predicate-expressions)
- [Predicate functions](#predicate-functions)
- [Packages](#packages)

### Pipe-forward operator
The **pipe-forward operator** (`|>`) sends output of a function as input to the next function.
In the [water treatment metaphor](/flux/v0.x/get-started/#flux-overview),

> the pipe-forward operator is that pipe the carries water or data through the pipeline.

Reviewer: ```suggestion
the pipe-forward operator is the pipe that carries water or data through the pipeline.
```

41 < 30
// Evaluates to false

"John" == "John" and 41 < 30
// Evaluates to false

"John" == "John" or 41 < 30
// Evaluates to true
```

Flux uses predicate expressions when [filtering data](#filter) or in
[conditional expressions](/flux/v0.x/spec/expressions/#conditional-expressions).

### Predicate functions
Predicate functions use [predicate expressions](#predicate-expressions) to evaluate
input and return `true` or `false`. For example:

```js
examplePredicate = (v) => v == "foo"

examplePredicate(v: "foo")
// Returns true

examplePredicate(v: "bar")
// Returns false
```

### Packages
The [Flux standard library](/flux/v0.x/stdlib/) is organized into [packages](/flux/v0.x/spec/packages/)
that contain functions and package-specific options.
The [universe package](/flux/v0.x/stdlib/universe/) is loaded by default.
To load other packages, include an import statement for each package at the
beginning of your Flux script.

```js
import "array"
import "math"
import "influxdata/influxdb/sample"
```

---

## Basic query structure

The majority of basic Flux queries include the following steps:

- [Source](#source)
- [Filter](#filter)
- [Shape](#shape)
- [Process](#process)

> ```js
> from(bucket: "example-bucket")            // ── Source
>   |> range(start: -1d)                    // ┐
>   |> filter(fn: (r) => r._field == "foo") // ┴─ Filter
>   |> group(columns: ["sensorID"])         // ── Shape
>   |> mean()                               // ── Process
> ```

Reviewer: ```suggestion
```js
from(bucket: "example-bucket")            // ── Source
  |> range(start: -1d)                    // ── Filter on time
  |> filter(fn: (r) => r._field == "foo") // ──  Filter on column values
  |> group(columns: ["sensorID"])         // ── Shape
  |> mean()                               // ── Process
```
```

  `filter()` uses a [predicate function](#predicate-functions) defined in the
  `fn` parameter to evaluate input rows.
  Each row is passed into the predicate function as a **record**, `r`, containing
  key-value pairs for each column in the row.

Other filter functions are also available.
For more information, see [Function types and categories – Filters](/flux/v0.x/function-types/#filters).

### Shape data
Many queries need to modify the structure of data to prepare it for processing.
Common data-shaping tasks include [regrouping data](/flux/v0.x/get-started/data-model/#restructure-tables)
by column values or by time or pivoting column values into rows.

Functions that reshape data include the following:

- [`group()`](/flux/v0.x/stdlib/universe/group/): modify group keys
- [`window()`](/flux/v0.x/stdlib/universe/window/): modify `_start` and `_stop` values of rows to group data by time
- [`pivot()`](/flux/v0.x/stdlib/universe/pivot/): pivot column values into rows
- [`drop()`](/flux/v0.x/stdlib/universe/drop/): drop specific columns
- [`keep()`](/flux/v0.x/stdlib/universe/keep/): keep specific columns and drop all others

### Process

> Processing data can take on many forms, but include the following types of operations:

Reviewer: ```suggestion
Processing data can take on many forms, and includes the following types of operations:
```

- **Select specific data points**: return specific rows from each input table.
  For example, return the first or last row, the row with the highest or lowest value, and more.
  For information, see [Function types and categories - Selectors](/flux/v0.x/function-types/#selectors).
- **Rewrite rows**: use [`map()`](/flux/v0.x/stdlib/universe/map/) to rewrite each input row.
  Tranform values with mathematic operations, process strings, dynamically add new columns, and more.
- **Send notifications**: evaluate data and use Flux notification endpoint functions
  to send notifications to external services. 
  For information, see [Function types and categories- Notification endpoints](/flux/v0.x/function-types/#notification-endpoints).

{{% note %}}
#### aggregateWindow helper function
[`aggregateWindow()`](/flux/v0.x/stdlib/universe/aggregatewindow/) is a helper function
that both **shapes and processes data**.
The function windows and groups data by time, and then applies an [aggregate](/flux/v0.x/function-types/#aggregates)
or [selector](/flux/v0.x/function-types/#selectors) function to the restructured tables.
{{% /note %}}

---

## Write a basic query

> Use [InfluxDB sample data](/{{< latest "influxdb" >}}/reference/sample-data/)to

Reviewer: ```suggestion
Use [InfluxDB sample data](/{{< latest "influxdb" >}}/reference/sample-data/) to
```

write a basic Flux query that queries data, filters the data by time and column values,

> then applies an [aggregate](/flux/v0.x/function-types/#aggregates).

Reviewer: ```suggestion
and then applies an [aggregate](/flux/v0.x/function-types/#aggregates).
```

{{% note %}}

> Use the [InfluxDB data explorer](/influxdb/cloud/query-data/execute-queries/data-explorer/)

Reviewer: ```suggestion
Use the [InfluxDB Data Explorer](/influxdb/cloud/query-data/execute-queries/data-explorer/)
```
